cns/restserver/ipam.go

```diff
@@ -891,3 +893,181 @@ func validateDesiredIPAddresses(desiredIPs []string) error {
+			}
+			return nil, errors.Wrap(err, "[GetEndpointState]  Failed to retrieve state")
+		} else if endpointInfo, ok := service.EndpointState[endpointID]; ok {
+			logger.Warnf("[GetEndpointState] Found existing endpoint state for container %s", endpointID)
+			return endpointInfo, nil
+		} else {
+			return nil, errors.New("endpoint could not be found in the statefile")
+		}
+	}
+	return nil, ErrOptManageEndpointState
+}
+
+// UpdateEndpointHandler handles the incoming UpdateEndpoint requests with http Patch method
+func (service *HTTPRestService) UpdateEndpointHandler(w http.ResponseWriter, r *http.Request) {
+	logger.Printf("[Azure CNS] updateEndpoint")
+
+	var req cns.EndpointRequest
+	err := service.Listener.Decode(w, r, &req)
+
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		response := cns.UpdateEndpointResponse{
+			Response: cns.Response{
+				ReturnCode: types.InvalidRequest,
+				Message:    fmt.Sprintf("[Azure CNS] updateEndpoint failed with error: %s", err.Error()),
+			},
+		}
+		err = service.Listener.Encode(w, &response)
+		logger.Response(service.Name, response, response.Response.ReturnCode, err)
+		return
+	}
+
+	err = service.UpdateEndpointHelper(req.EndpointID, req.HnsID, req.VethName)
+	if err != nil {
+		response := cns.UpdateEndpointResponse{
+			Response: cns.Response{
+				ReturnCode: types.UnexpectedError,
+				Message:    fmt.Sprintf("[Azure CNS] updateEndpoint failed with error: %s", err.Error()),
+			},
+		}
+		err = service.Listener.Encode(w, &response)
+		logger.Response(service.Name, response, response.Response.ReturnCode, err)
+	} else {
+		response := cns.UpdateEndpointResponse{
+			Response: cns.Response{
+				ReturnCode: types.Success,
+				Message:    "[Azure CNS] updateEndpoint retruned successfully",
+			},
+		}
+		err = service.Listener.Encode(w, &response)
+		logger.Response(service.Name, response, response.Response.ReturnCode, err)
+	}
+}
+
+// UpdateEndpointHelper updates the state of the given endpointId with HNSId or VethName
+func (service *HTTPRestService) UpdateEndpointHelper(endpointID, hnsID, vethName string) error {
+	if service.EndpointStateStore == nil {
+		return ErrStoreEmpty
+	}
+	logger.Printf("[UpdateEndpointState] Updating endpoint state for infra container %s", endpointID)
+	if service.Options[common.OptManageEndpointState] == true {
+		if endpointInfo, ok := service.EndpointState[endpointID]; ok {
+			logger.Printf("[UpdateEndpointState] Found existing endpoint state for infra container %s", endpointID)
+			if hnsID != "" {
+				service.EndpointState[endpointID].HnsID = hnsID
+				logger.Errorf("[UpdateEndpointState] update the endpoint %s with HNSID  %s", endpointID, hnsID)
+				return nil
+			} else if vethName != "" {
+				service.EndpointState[endpointID].VethName = vethName
+				logger.Errorf("[UpdateEndpointState] update the endpoint %s with vethName  %s", endpointID, vethName)
+				return nil
+			}
```

Comment:
convert to if

Reply:
I put else if in here to make it impossible for the client to put both vethname and hnsID in a statefile since we only have one of them for Linux and Windows.